File: src/main/java/com/zzc/weekcompetition/LengthOfLongestSubsequence.java
package com.zzc.weekcompetition;

import java.util.*;

/**
 * @author zc.zhou
 * @Description TODO: 01 背包 100042. 和为目标值的最长子序列的长度
 * @create 2023-10-28 22:48
 */
public class LengthOfLongestSubsequence {
    public static void main(String[] args) {
        LengthOfLongestSubsequence l = new LengthOfLongestSubsequence();
        //[2,13,7,3,14,8,11,10,7,13]
        //44
        List<Integer> list = new ArrayList<>(List.of(2,13,7,3,14,8,11,10,7,13));
        //List<Integer> list = new ArrayList<>(List.of(1,1,5,4,5));
        System.out.println(l.maxSubArrayLen(list, 44));
    }
    public int maxSubArrayLen(List<Integer> nums, int target) {
        int[] res = new int[target + 1];
        Arrays.fill(res, -1);
        res[0] = 0;
        for (int num : nums) {
            for (int i = target; i >= num; i--) {
                if (res[i - num] != -1) {
                    res[i] = Math.max(res[i], res[i - num] + 1);
                }
            }
        }
        return res[target];
    }

}

File: pyproject/dp/knapsack/LengthOfLongestSubsequence.py
from typing import List


class Solution:
    # 2915和为目标值的最长子序列的长度
    def lengthOfLongestSubsequence(self, nums: List[int], target: int) -> int:
        size = len(nums)
        # dp[i][j]: 表示nums[0] 到nums[i] 和为j的最长子序列的长度
        dp = []
        for i in range(size):
            dp.append([-1] * (target + 1))
        if nums[0] <= target:
            dp[0][nums[0]] = 1
        dp[0][0] = 0
        for i in range(1, size):
            # 考虑第i个数选或者不选
            for j in range(target, -1, -1):
                # 不选第i个数
                dp[i][j] = dp[i - 1][j]
                if j - nums[i] >= 0 and dp[i - 1][j - nums[i]] != -1:
                    # 不选: dp[i-1][j] 选:dp[i-1][j-num[i]]
                    dp[i][j] = max(dp[i - 1][j], dp[i - 1][j - nums[i]] + 1)
        return dp[size - 1][target]


s = Solution()
print(s.lengthOfLongestSubsequence([1, 3, 3, 8], 7))
